### PR TITLE
Reactions to messages

### PR DESCRIPTION
**Pull request by hildjj**
React and unreact to messages, list the reactions for a message.
TODO: list reactions on a User.

As discussed in feedback+id525916.

### CODE REVIEW

**Comment by josephpage** [comment]
:+1:

**Comment by visnup** [comment]
:+1:

**Comment by Connoropolous** [comment]
Is there a plan/timeline for merging this in?

**Comment by albatrocity** [comment]
:+1: this would be handy

File: src/message.coffee [hunk 1]
```diff
@@ -72,18 +72,69 @@ class Message
...
-    }
+    if @type == 'reaction_added'
+      util = require 'util'
+      console.log JSON.stringify(@)
+      if @item?
+        params =
+          timestamp: @item.ts
+          channel: @item.channel
+          name: @reaction
+        @_client.logger.debug "Sending reation delete request"
```
Comment (msabramo) [inline src/message.coffee]: reation => reaction

**Comment by msabramo** [comment]
This is pretty awesome! Thanks!

One thing that would be nice is if the `reaction_added` event could give me more information about the message that the reaction was added for.

Currently, it gives me data like:

```json
{ type: 'reaction_added',
  user: 'U02AXCUAJ',
  item:
   { type: 'message',
     channel: 'C0CA0D74M',
     ts: '1446213270.001126' },
  reaction: 'simple_smile',
  event_ts: '1446217125.132010' }
```

but I don't know what the message text is for the item that the reaction was to. Is it possible for it to pass the message text to the callback?

**Comment by msabramo** [comment]
Anyone from Slack looking at this yet? @l12s perhaps?

**Comment by l12s** [comment]
Hi there,

So, I'm a little confused on this one. I'd have expected the `item` object to have a `message` property that looks like https://api.slack.com/events/message

I'll try and take a look into what the response is early next week.

Leah

**Comment by msabramo** [comment]
I'm playing around with using this:

```coffeescript
      when 'reaction_added'
        channel = @getChannelGroupOrDMByID message.item.channel
        if channel._history[message.item.ts]
          message.item = channel._history[message.item.ts]
        m = new Message @, message
        @emit 'reaction_added', m
```

This is nice because then I can do:

```coffeescript
slack.on 'reaction_added', (message) ->
    channel = message._client.channels[message.item.channel].name
    console.log(
      "Reaction added: message.reaction=#{message.reaction} " +
      "channel=\"#{channel}\" " +
      "message.item.text=\"#{message.item.text}\"")
```

and get:

```
Message received: channel="test_devopsbot" user="marca" text="Hello world!"
Reaction added: message.reaction=simple_smile channel="test_devopsbot" message.item.text="Hello world!"
```

**Comment by msabramo** [comment]
I'd love to get some feedback on the approach that I took if someone has time to take a look.

I have a forked/hacked environment that I've built a little upon and I don't want to build too much on it if the foundation is going to change.

**Comment by l12s** [comment]
Sorry, I'll take a look tomorrow!

**Comment by l12s** [comment]
Sorry about the delay in getting to this today. We're in the process of wrapping up a proposed refactor that would address this issue in a more fundamental way by separating the caching logic from the client layer, so you could do the work you're proposing above to get the message in an event handler.

I'm hoping to wrap up that proposed refactor this weekend and have a changelog + commits ready for comments and review within a couple weeks. So, if you can wait a while, I'd suggest doing so.

**Comment by CaptainPalapa** [comment]
*We* appreciate y'all's work!

**Comment by msabramo** [comment]
OK -- thanks, @l12s! Yeah that's what I figured -- that what I was doing was probably not the best approach and that you folks could probably do it in a better way. I will wait a bit for the nicer solution.

Thanks!

**Comment by l12s** [comment]
Ok, first off, I apologize for how long it's taken us to resolve this PR.

I just pushed the 2.0.0 release for this library, which is a ground up refactor with the aim of building a lib that both Slack and our external contributors can easily work with and improve.

That release now passes through all messages from the RTM API by default as well as [providing handlers for reactions](https://github.com/slackhq/node-slack-client/blob/03743068caa535f15a636da164953f402d1899c5/lib/data-store/message-handlers/reactions.js) against objects in the in-memory data store. Hopefully that gives you what you need, but if not, please raise an issue or create a new PR.

**Comment by hildjj** [comment]
Are you already working with the Hubot folks to transition to the new version?

**Comment by l12s** [comment]
Not yet - my rough plan on this is to have the `>= 2.0.x` release out for 2-4 weeks to shake out any bugs and then work with botkit and hubot to transition them over. I'll create an issue to track that work.

**Pull request closed without merging**